### ISSUE TITLE
Typeful API in `kotools.csv`

### ISSUE THREAD

**Issue by LVMVRQUXL** [kotools/libraries#14]
## Description

Create a package `kotools.csv` containing all declarations of the package `io.github.kotools.csv`.
Then, deprecate all declarations of the old package.
> Declarations in `kotools.csv` should use explicit types from Kotools Types for having a typeful design.

Also, the `file` property should be required in this new API for avoiding runtime checks in favor of compile-time checks.
Here is an exemple using the old API:
```kotlin
data class Person(val name: NotBlankString, val age: StrictlyPositiveInt, val isAdmin: Boolean = false)

suspend fun main() {
    csvWriter<Person>(file = "people".toNotBlankString()) {
        records { +Person("Nobody".toNotBlankString(), 25.toStrictlyPositiveInt()) }
    }
    val people: List<Person> = csvReader(file = "people".toNotBlankString())
    println(people)
}
```

## Checklist

- [ ] Implement a basic CSV reader.
- [ ] Implement a reader that returns a custom type.
- [ ] Deprecate the old reader.
- [ ] Implement a basic CSV writer.
- [ ] Implement a writer that takes a custom type.
- [ ] Deprecate the old writer.
- [ ] Update `Work in progress` section in changelog.

**Comment by LVMVRQUXL** [comment]
The issue of the last example is that unwrapping the `CsvReaderResult` type could generate boilerplate code...
A solution could be to return an `Either<CsvReaderError, NotEmptyList<Record>>` type instead, where the `Either` type and its operations are provided by [Arrow Core](https://github.com/arrow-kt/arrow).

**Comment by LVMVRQUXL** [comment]
> The issue of the last example is that unwrapping the `CsvReaderResult` type could generate boilerplate code... A solution could be to return an `Either<CsvReaderError, NotEmptyList<Record>>` type instead, where the `Either` type and its operations are provided by [Arrow Core](https://github.com/arrow-kt/arrow).
> 
> Also, the API should use explicit types only for reducing the amount of runtime checks: the `file` parameter should be a `NotBlankString`.

We will not use the `Either` type provided by [Arrow Core](https://github.com/arrow-kt/arrow). Instead, we will introduce a new `CsvReaderResult` type like proposed in [the previous example](https://github.com/kotools/libraries/issues/14#issuecomment-1304066389).

**Comment by LVMVRQUXL** [comment]
Will be implemented by #45 and #46.

**Comment by LVMVRQUXL** [comment]
Reading a CSV file is an IO operation that should run inside a coroutine.
Here's a new example of the API:
```kotlin
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotools.types.string.NotBlankString
import kotools.types.string.notBlankStringOrThrow

/**
 * Returns the [records][CsvReaderResult.Success.records] in the CSV file
 * matching this path, or returns a [CsvReaderResult.Exception.FileNotFound] if
 * no file matches this path, or returns a [CsvReaderResult.Exception.EmptyFile]
 * if the file matching this path is empty.
 */
internal suspend fun CsvPathResult.Success.read(): CsvReaderResult =
    withContext(CoroutineName("CsvReader") + Dispatchers.IO) { TODO() }

internal sealed interface CsvReaderResult {
    class Success(val records: List<Map<String, String>>) : CsvReaderResult

    sealed class Exception(reason: NotBlankString) :
        IllegalStateException("The file ${reason.value}."),
        CsvReaderResult {
        class EmptyFile(file: NotBlankString) : Exception(
            notBlankStringOrThrow("$file is empty")
        )

        class FileNotFound(file: NotBlankString) : Exception(
            notBlankStringOrThrow("$file doesn't exist")
        )
    }
}
```
